server/src/main/java/org/opensearch/cluster/routing/RoutingNodes.java

```diff
@@ -1378,32 +1381,89 @@ public void remove() {
                 }
             };
         } else {
-            return new Iterator<ShardRouting>() {
-                @Override
-                public boolean hasNext() {
-                    while (!queue.isEmpty()) {
-                        if (queue.peek().hasNext()) {
+            if (shardMovementStrategy == BalancedShardsAllocator.ShardMovementStrategy.REPLICA_FIRST) {
```

Comment:
Do we need to add an explicit check in `NodeVersionAllocationDecider` that restricts a primary to get allocated to a newer version node if it has at least one replica in an older version. 
Also for cases where a replica that moves to a newer version node first, can get eligible for promotion to primary(based on target node version) if an existing primary fails, even though other replicas might be on older version. Do we need to handle those cases as well?

Reply:
+1 for using `NodeVersionAllocationDecider` for restricting primary shard allocations. Gating this logic only for segment replication enabled indices will make it less intrusive compared to prioritizing moving all replica shards first. 

>Also for cases where a replica that moves to a newer version node first, can get eligible for promotion to primary(based on target node version) if an existing primary fails, even though other replicas might be on older version. Do we need to handle those cases as well?

Good catch! Yes, this needs to be handled separately. Today, we promote the active replica running on [highest](https://github.com/opensearch-project/OpenSearch/blob/a3aab67ee86bf171a7eb480f0933e0b955fbf4f3/server/src/main/java/org/opensearch/cluster/routing/RoutingNodes.java#L375) OpenSearch version. We can promote replica running on older version for segment replication enabled indices.

Reply:
@Poojita-Raj: I think all we need is update in NodeVersionAllocationDecider logic to prevent primary shard allocation (canAllocate) when there are replicas on older version nodes. We don't need to force move replicas first. The change in `NodeVersionAllocationDecider` is sufficient and achieves replicas first movement. @Bukhtawar : Let know if we are missing something.

Reply:
Apart from this, like @Bukhtawar called out, we'll also need to handle various replica to primary promotion cases when cluster is in mixed mode.

Reply:
Also, regarding the approach of relying on `NodeVersionAllocationDecider`, what would be the exact behaviour during b/g or rolling upgrade? Would primary simply fail to relocate to newer version(with `NodeVersionAllocationDecider ` as 'No'), and then with retry, move after all replicas are on newer version?

Reply:
Thanks @ankitkala for the comment. 

> Apart from this, like @Bukhtawar called out, we'll also need to handle various replica to primary promotion cases when cluster is in mixed mode.

I think the only case we need to handle is failover during mixed cluster mode, which will be solved by promoting replica running on older OpenSearch version. Are there more use-cases we need to consider here. If there are more use-cases, can you please share ? 

>Also, regarding the approach of relying on NodeVersionAllocationDecider, what would be the exact behaviour during b/g or rolling upgrade? 

For both b/g (relocation based upgrades) and rolling upgrade, any attempt to allocate primary shard will fail when any replica shard is still on older version. 

> Would primary simply fail to relocate to newer version(with NodeVersionAllocationDecider  as 'No'), and then with retry, move after all replicas are on newer version?

Primary relocation wouldn't be attempted as `NodeVersionAllocationDecider` decider will continue to return `NO` till all replicas are on higher version.

Reply:
Created #9027 to track failover in mixed version clusters.

Reply:
> I think the only case we need to handle is failover during mixed cluster mode, which will be solved by promoting replica running on older OpenSearch version.

Yep. I think that's the only one. 

> Primary relocation wouldn't be attempted as NodeVersionAllocationDecider decider will continue to return NO till all replicas are on higher version.

Got it. So after all replicas are on new version, the next rebalance call will trigger the primary relocation.

Reply:
Replicas relocation to higher OpenSearch version nodes would trigger cluster state update event which results in follow up [reroute](https://github.com/opensearch-project/OpenSearch/blob/d916f9c1027f5b2ccff971a66921b7c26db1688f/server/src/main/java/org/opensearch/cluster/routing/allocation/AllocationService.java#L493). Reroute would then go through steps of allocateUnassigned, moveShards, rebalance. When all replicas are relocated, the primary relocation would be attempted inside moveShards step. All relocations are triggered inside moveShards since we are setting exclude filters.